libs/SalesforceSDKCore/SalesforceSDKCoreTests/SFUserAccountManagerPersisterTests.m

```diff
@@ -0,0 +1,381 @@
+ with the distribution.
+ * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
+ endorse or promote products derived from this software without specific prior written
+ permission of salesforce.com, inc.
+
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <XCTest/XCTest.h>
+#import <SalesforceSDKCore/SalesforceSDKCore.h>
+#import "SFDefaultUserAccountPersister.h"
+#import "SFUserAccountPersisterEphemeral.h"
+#import "SFUserAccountManager+Internal.h"
+
+static NSString * const kUserIdFormatString = @"005R0000000Dsl%lu";
+static NSString * const kOrgIdFormatString = @"00D000000000062EA%lu";
+
+@interface SFUserAccountManagerPersisterTests : XCTestCase
+@property (nonatomic, strong) SFUserAccountManager *uam;
+@end
+
+@interface TestUserAccountManagerPersisterDelegate : NSObject <SFUserAccountManagerDelegate>
+
+@property (nonatomic, strong) SFUserAccount *willSwitchOrigUserAccount;
+@property (nonatomic, strong) SFUserAccount *willSwitchNewUserAccount;
+@property (nonatomic, strong) SFUserAccount *didSwitchOrigUserAccount;
+@property (nonatomic, strong) SFUserAccount *didSwitchNewUserAccount;
+
+@end
+
+@implementation TestUserAccountManagerPersisterDelegate
+
+- (id)init {
+    self = [super init];
+    if (self) {
+        [[SFUserAccountManager sharedInstance] addDelegate:self];
+    }
+    return self;
+}
+
+- (void)dealloc {
+    [[SFUserAccountManager sharedInstance] removeDelegate:self];
+}
+
+- (void)userAccountManager:(SFUserAccountManager *)userAccountManager
+        willSwitchFromUser:(SFUserAccount *)fromUser
+                    toUser:(SFUserAccount *)toUser {
+    self.willSwitchOrigUserAccount = fromUser;
+    self.willSwitchNewUserAccount = toUser;
+}
+
+- (void)userAccountManager:(SFUserAccountManager *)userAccountManager
+         didSwitchFromUser:(SFUserAccount *)fromUser
+                    toUser:(SFUserAccount *)toUser {
+    self.didSwitchOrigUserAccount = fromUser;
+    self.didSwitchNewUserAccount = toUser;
+}
+
+@end
+
+@implementation SFUserAccountManagerPersisterTests
+
+- (void)setUp {
+    [super setUp];
+    self.uam = [SFUserAccountManager sharedInstance];
+    self.uam.accountPersister = [SFUserAccountPersisterEphemeral new];
```

Comment:
This needs to either be a) tested against an `SFUserAccountManager` instance in isolation, or b) more of a temporary-change-then-replacement pattern.  Because in the latter case, when running in the context of other tests, we're changing the account persister somewhere in the middle, which can have unintended effects on full test run behavior.

The pattern I usually use is:

- In `setUp`, persist the current account persister value in a placeholder property (e.g. `self.origAccountPersister`), then set the new value for `self.uam.accountPersister` as desired.
- In `tearDown`, reset `self.uam.accountPersister` to `self.origAccountPersister`.